tuned/plugins/plugin_irq.py

```diff
@@ -0,0 +1,113 @@
+from . import hotplug
+from .decorators import *
+import tuned.logs
+from tuned.utils.commands import commands
+
+import errno
+import os
+
+log = tuned.logs.get()
+
+class IrqPlugin(hotplug.Plugin):
+
+	def __init__(self, monitor_repository, storage_factory, hardware_inventory, device_matcher, device_matcher_udev, plugin_instance_factory, global_cfg, variables):
+		super(IrqPlugin, self).__init__(monitor_repository, storage_factory, hardware_inventory, device_matcher, device_matcher_udev, plugin_instance_factory, global_cfg, variables)
+
+	def _init_devices(self):
+		self._devices_supported = True
+		self._free_devices = set()
+		self._assigned_devices = set()
+		# print([d.sys_name for d in self._hardware_inventory.get_devices("net")])
+		for i in os.listdir("/proc/irq"):
+			p = os.path.join("/proc/irq", i)
+			if os.path.isdir(p) and i.isdigit():
+				self._free_devices.add("irq%s" % i)
+
+	@classmethod
+	def _get_config_options(self):
+		return {
+			"affinity": "",
+		}
+
+	def _instance_init(self, instance):
+		log.info(f"_instance_init({instance.name})")
+		instance._has_static_tuning = True
+		instance._has_dynamic_tuning = True
+		instance._irq_original = {}
+		instance._irq_unchangeable = []
+		instance._is_main_instance = instance.name == instance.plugin.name
+		instance._created_dynamically = False
+
+		affinity = instance.options.get("affinity")
+		affinity_list = self._cmd.cpulist_unpack(affinity)
+		instance._affinity = affinity_list
+		if len(affinity_list) == 0:
+			log.info(f"instance {instance.name} has no affinity. deactivating.")
+			instance._active = False
+
+	def _instance_cleanup(self, instance):
+		log.info(f"_instance_cleanup({instance.name})")
+
+	def _instance_apply_static(self, instance):
+		log.info(f"_instance_apply_static({instance.name})")
+
+	def _instance_unapply_static(self, instance, rollback):
+		log.info(f"_instance_unapply_static({instance.name})")
+
+	def _instance_apply_dynamic(self, instance, device):
+		log.info(f"_instance_apply_dynamic({instance.name}, {device})")
```

Comment:
This requires dynamic tuning enabled, which is by default disabled on RHEL. I think it should go under the static method. Or maybe better to implement it in the `command_custom` .

You could also implement it in the `command_set` /  `command_get` which will be called for each device/irq. It's probably the most clean solution, but it could have some overhead. Taking into account that on typical system there wouldn't be usually much IRQs, this could work.

Reply:
I believe you should call
```python
super(IrqPlugin, self)._instance_apply_static(instance)
```
at the beginning of `IrqPlugin._instance_apply_static` to execute all decorated commands (in this case, the custom affinity command). Then, you don't have to manually loop through the list of assigned devices.